src/main/java/com/marklogic/spark/reader/file/MlcpArchiveFileReader.java

```diff
@@ -1,61 +1,166 @@
+        } catch (Exception e) {
+            String message = String.format("Unable to read metadata for entry: %s; file: %s; cause: %s",
+                metadataZipEntry.getName(), this.path, e.getMessage());
+            if (fileContext.isReadAbortOnFailure()) {
+                throw new ConnectorException(message, e);
+            }
+            // Contrary to a zip of aggregate XML files, if we get a bad metadata file, we likely do not have a valid
+            // MLCP archive file and thus there's no reason to continue processing this particular file.
+            Util.MAIN_LOGGER.warn(message);
+            return null;
+        }
+    }
+
+    private ZipEntry getContentEntry(ZipEntry metadataZipEntry) {
+        ZipEntry contentZipEntry;
+        try {
+            contentZipEntry = FileUtil.findNextFileEntry(zipInputStream);
         } catch (IOException e) {
-            throw new ConnectorException(String.format("Unable to read from zip file at %s; cause: %s",
-                this.path, e.getMessage()), e);
+            String message = String.format("Unable to read content entry from file: %s; cause: %s", this.path, e.getMessage());
+            if (fileContext.isReadAbortOnFailure()) {
+                throw new ConnectorException(message, e);
+            }
+            Util.MAIN_LOGGER.warn(message);
+            return null;
+        }
+
+        if (contentZipEntry == null) {
+            String message = String.format("No content entry found for metadata entry: %s; file: %s", metadataZipEntry.getName(), this.path);
+            if (fileContext.isReadAbortOnFailure()) {
+                throw new ConnectorException(message);
+            }
+            Util.MAIN_LOGGER.warn(message);
+            return null;
         }
-        long length = content.length;
-        return new GenericInternalRow(new Object[]{
-            UTF8String.fromString(uri), null, length, ByteArray.concat(content)
-        });
+        return contentZipEntry;
     }
 
-    @Override
-    public void close() throws IOException {
-        this.zipInputStream.close();
+    private byte[] readBytesFromContentEntry(ZipEntry contentZipEntry) {
+        try {
+            return FileUtil.readBytes(zipInputStream);
+        } catch (IOException e) {
+            String message = String.format("Unable to read entry %s from zip file at %s; cause: %s",
+                contentZipEntry.getName(), this.path, e.getMessage());
+            if (fileContext.isReadAbortOnFailure()) {
+                throw new ConnectorException(message);
+            }
+            Util.MAIN_LOGGER.warn(message);
+            return new byte[0];
+        }
+    }
+
+    private InternalRow makeRow(ZipEntry contentZipEntry, byte[] content, MlcpMetadata mlcpMetadata) {
+        final String zipEntryName = contentZipEntry.getName();
+        final String uri = zipEntryName.startsWith("/") ?
```

Comment:
Can we can the private functions to a utilities class and make it accessible by other classes in the same package? These might be needed for ArchiveFileReader as well.

Reply:
Definitely! But let's wait until we need it. I always like to duplicate the code first, get tests in place, and make sure it's working. Then we can do a follow-up PR that is only refactoring that gets rid of the duplication, as we have the tests to ensure we don't break anything in the process.